### PR TITLE
[TensorExpr] Add aten::sum lowering to the kernel

### PR DESCRIPTION
**Pull request by asuhan**
Summary:
Handles all dimensions and selected dimensions, per PyTorch semantics.

Test Plan: test_tensorexpr

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by nickgg — COMMENTED** [review]
To disable vectorization for reductions this is what I'd do:
```
1374: if (NodeFinder<ReduceOp>::find(loop).size() > 0) {
1375:     continue;
1376: }
```

not backseating just trying to help!

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1591,6 +1593,106 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
   }
 }
 
+namespace {
+
+std::vector<VarHandle> squeezeIndices(
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1604]: A comment here would help a lot in understanding this helper.

**Review by nickgg — COMMENTED** [review]
lgtm!

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 2]
```diff
@@ -261,5 +262,164 @@ void testKernel_4() {
...
+      sizes.begin(), sizes.end(), 1, std::multiplies<int64_t>());
+  std::vector<float> values(numel);
+  std::iota(values.begin(), values.end(), 0);
+  auto a = at::tensor(values, options);
+  return a.reshape(sizes);
+}
+
+} // namespace
+
+void testKernel_5() {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply test/cpp/tensorexpr/test_kernel.cpp]: Done.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1605]: Done.

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 3]
```diff
@@ -261,5 +262,164 @@ void testKernel_4() {
...
+
+void testKernelSumOneAxis() {
+  // Test lowering of sum on one axis.
+  const auto graph_template = R"IR(
+      graph(%0 : Float(5:3,3:1, device=cpu)):
+        %1 : int = prim::Constant[value=${dim}]()
+        %2 : int[] = prim::ListConstruct(%1)
+        %3 : bool = prim::Constant[value=${keepdim}]()
+        %4 : ${dtype}
+        %5 : Float() = aten::sum(%0, %2, %3, %4)
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_kernel.cpp:333]: The output shape seems to be wrong if we're summing only one axis (and it will also depend on the `${keepdim}`. You could use `Tensor` here, but that would require adding supporting `aten::sum` in the NNC shape inference (which shouldn't be very difficult).

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 4]
```diff
@@ -261,5 +262,164 @@ void testKernel_4() {
...
+void testKernelSumMultipleAxes() {
+  // Test lowering of sum on multiple axes.
+  const auto graph_template = R"IR(
+      graph(%0 : Float(5:72,3:24,4:6,6:1, device=cpu)):
+        %1 : int = prim::Constant[value=${dim1}]()
+        %2 : int = prim::Constant[value=${dim2}]()
+        %3 : int[] = prim::ListConstruct(%1, %2)
+        %4 : bool = prim::Constant[value=${keepdim}]()
+        %5 : ${dtype}
+        %6 : Float() = aten::sum(%0, %3, %4, %5)
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_kernel.cpp:380]: Ditto.

**Comment by asuhan** [comment]
@nickgg I had to move the check before the reduction expander runs, otherwise we wouldn't find a `ReduceOp`.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 5]
```diff
@@ -1591,6 +1602,107 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
...
+        }
+      },
+      reduction_info.reductionDims);
+}
+
+TensorExprKernel::ReductionInfo TensorExprKernel::getReductionInfo(
+    const torch::jit::Node* node) {
+  std::vector<size_t> axes;
+  bool keepdim = false;
+  auto sizes = sizesForValue(node->namedInput(attr::self));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1662]: Hm, that's a little surprising, what IR do you expect to see here? When I'm inspecting IR for `x.sum()` it doesn't have any attributes:
```
import torch

def foo(x):
    return x.sum()

f = torch.jit.script(foo)
print(f.graph)
```
The corresponding IR:
```
graph(%x.1 : Tensor):
  %2 : None = prim::Constant()
  %3 : Tensor = aten::sum(%x.1, %2) # u.py:4:11
  return (%3)
```

What node are we calling `namedInput` on?

In either case, I think it could be useful to show the expected IR in a comment.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 6]
```diff
@@ -1367,6 +1374,10 @@ Stmt* TensorExprKernel::generateStmt(BackendType backendType) {
 
     // vectorize inner loops.
     for (For* loop : innerLoops) {
+      if (NodeFinder<ReduceOp>::find(loop).size() > 0) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1377]: Is it needed given the `allowVectorization` check above?

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 5 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1662]: Actually, nvm, I didn't know we had such a fancy API for accessing inputs by their name in schema :) Showing the schema here could still be helpful.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 7]
```diff
@@ -1591,6 +1602,107 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
...
+TensorExprKernel::ReductionInfo TensorExprKernel::getReductionInfo(
+    const torch::jit::Node* node) {
+  std::vector<size_t> axes;
+  bool keepdim = false;
+  auto sizes = sizesForValue(node->namedInput(attr::self));
+  const auto inputs = node->inputs();
+  if (inputs.size() > 2) {
+    // Canonicalize axes: wrap around, sort and make unique.
+    auto axesValue = node->namedInput(attr::dim);
+    for (auto axisNode : axesValue->node()->inputs()) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1667]: I think we're expecting to see `prim::ListConstruct` here and we're iterating over its inputs, which are expected to be `prim::Constant`s? Could we add some asserts to actually verify that?

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 8]
```diff
@@ -261,5 +262,164 @@ void testKernel_4() {
...
+              IntArrayRef{dim1, dim2}, /*keepdim=*/keepdim, /*dtype=*/dtype);
+          TensorExprKernel k(graph);
+          std::vector<at::Tensor> inputs = {a};
+          Stmt* s = k.getCodeGenStmt();
+          // TODO: verify stmt
+
+          std::vector<IValue> stack = fmap<IValue>(inputs);
+          k.run(stack);
+          o = stack[0].toTensor();
+          ASSERT_EQ(o.sizes(), ref.sizes());
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_kernel.cpp:416]: Should we be checking dtypes as well?

**Review by ZolotukhinM — APPROVED** [review]
Very nice! I have only minor comments but overall looks good!

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 3 above)
Comment (asuhan) [reply test/cpp/tensorexpr/test_kernel.cpp]: Just `Tensor` worked, probably because we don't have fusion with any other operator in this PR. I'll have to add the shape inference in my next PRs, yes.

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 4 above)
Comment (asuhan) [reply test/cpp/tensorexpr/test_kernel.cpp]: Done, `Tensor` worked.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 5 above)
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1659]: I've added a comment, not sure what's the confusion here or whether the expected IR would improve matters. The `aten::sum`  operator has a named input `self`, per schema: https://github.com/pytorch/pytorch/blob/0521c712410ae014844d680cfc40b59c6f456bcf/aten/src/ATen/native/native_functions.yaml#L2939. Is there a different / better way to do this? The `attr::` namespace is a bit confusing, since it's not an actual attribute.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 6 above)
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Oops, good catch! Done.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 7 above)
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1665]: I've added an assert to check it's `prim::ListConstruct`. Subsequent casts of elements to `IntImm` should take care of asserting the rest.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 5 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1659]: I usually just referenced inputs by their index (i.e. `n->input(2)`) - but I think addressing them by name makes it a bit easier to read despite they are not really attributes.

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 8 above)
Comment (asuhan) [reply test/cpp/tensorexpr/test_kernel.cpp:418]: Done. Meant to do it and forgot about it, good catch.

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 3 above)
Comment (ZolotukhinM) [reply test/cpp/tensorexpr/test_kernel.cpp]: Yeah, it probably would've failed if we tried to use the output as an input of some other node.

File: torch/csrc/jit/tensorexpr/kernel.h [hunk 9]
```diff
@@ -129,6 +131,18 @@ class TORCH_API TensorExprKernel {
...
+    std::vector<DimArg> reductionDims;
+    std::vector<DimArg> outputDims;
+    std::vector<size_t> axes;
+    bool keepdim;
+    c10::optional<Dtype> dtype;
+  };
+
+  // Get the reduction info for the given node, based on properties and inputs.
+  ReductionInfo getReductionInfo(const torch::jit::Node* node);
+
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/kernel.h:145]: Do these need to be exposed in the header or can they be local to kernel.cpp?

**Review by bertmaher — APPROVED** [review]

File: torch/csrc/jit/tensorexpr/kernel.h (hunk 9 above)
Comment (asuhan) [reply torch/csrc/jit/tensorexpr/kernel.h:145]: Yes, unfortunately. The `getReductionInfo` method needs to be a non-static member of `TensorExprKernel` because it calls `sizesForValue`.
